File: scripts/split.js
/**
 * One-time migration: split Model-Database.json into src/{type}/{category}-{number}-{name}.json
 * Type folders and filenames are lowercased; spaces become dashes.
 */

const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const SOURCE = path.join(ROOT, 'Model-Database.json');
const SRC_DIR = path.join(ROOT, 'src');

/**
 * Derive type folder name: lowercase, spaces → dashes.
 */
function typeToFolder(type) {
  if (type == null || type === '') return 'unknown';
  return String(type)
    .toLowerCase()
    .replace(/\s+/g, '-')
    .replace(/[\\/:*?"<>|]/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '') || 'unknown';
}

/**
 * Sanitize for filename: lowercase, replace spaces and invalid chars with dash,
 * strip Unicode symbols (™, ®), collapse dashes, trim.
 */
function sanitize(str) {
  if (str == null || str === '') return 'unknown';
  let s = String(str)
    .toLowerCase()
    .replace(/[\s\\/:*?"<>|]/g, '-')
    .replace(/[\u2122\u00AE]/g, '') // ™ ®
    .replace(/[^\p{L}\p{N}-]/gu, '-') // other non-letters/numbers/dash → dash
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');
  return s || 'unknown';
}

/**
 * Derive filename: {category}-{number}-{name}.json (all lowercase, no spaces).
 */
function getFilename(entry) {
  const type = entry.type;
  const category = entry.category;
  const number = entry.number;
  const name = entry.name;

  const typeFolder = typeToFolder(type);
  const cat = sanitize(category);
  const num = number != null ? sanitize(String(number)) : 'unknown';
  const n = sanitize(name);

  return `${cat}-${num}-${n}.json`;
}

function main() {
  if (!fs.existsSync(SOURCE)) {
    console.error('Source file not found:', SOURCE);
    process.exit(1);
  }

  const data = JSON.parse(fs.readFileSync(SOURCE, 'utf8'));
  if (!Array.isArray(data)) {
    console.error('Expected JSON array');
    process.exit(1);
  }

  const seen = new Map(); // key = dir + base filename → count for collision suffix

  let skipped = 0;
  for (let i = 0; i < data.length; i++) {
    const entry = data[i];
    const type = entry.type;
    const typeFolder = typeToFolder(type);
    if (!type || !entry.category || entry.category === '' || !entry.number || entry.number === '' || !entry.name || entry.name === '') {
      console.warn(`[${i}] Skipping entry missing type/category/number/name:`, entry.number || entry.name || '(no id)');
      skipped++;
      continue;
    }

    const dir = path.join(SRC_DIR, typeFolder);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    let filename = getFilename(entry);
    const key = path.join(dir, filename);
    let count = seen.get(key) || 0;
    seen.set(key, count + 1);
    if (count > 0) {
      const ext = path.extname(filename);
      const base = path.basename(filename, ext);
      filename = `${base}-${count + 1}${ext}`;
    }

    const outPath = path.join(dir, filename);
    fs.writeFileSync(outPath, JSON.stringify(entry, null, 4), 'utf8');
  }

  console.log(`Wrote ${data.length - skipped} entries to ${SRC_DIR}`);
  if (skipped > 0) console.log(`Skipped ${skipped} entries.`);
}

main();

File: scripts/split.py
"""
One-time migration: split Model-Database.json into src/{type}/{category}-{number}-{name}.json
Type folders and filenames are lowercased; spaces become dashes.
"""

import json
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SOURCE = ROOT / "Model-Database.json"
SRC_DIR = ROOT / "src"


def type_to_folder(type_val):
    """Derive type folder name: lowercase, spaces -> dashes."""
    if type_val is None or type_val == "":
        return "unknown"
    s = str(type_val).lower().replace(" ", "-")
    s = re.sub(r'[\\/:*?"<>|]', "-", s)
    s = re.sub(r"-+", "-", s).strip("-")
    return s or "unknown"


def sanitize(s):
    """Sanitize for filename: lowercase, replace spaces and invalid chars with dash, strip ™ ®, collapse dashes."""
    if s is None or s == "":
        return "unknown"
    s = str(s).lower()
    s = re.sub(r'[\s\\/:*?"<>|]', "-", s)
    s = s.replace("\u2122", "").replace("\u00ae", "")  # ™ ®
    s = re.sub(r"[^\w-]", "-", s, flags=re.UNICODE)
    s = re.sub(r"-+", "-", s).strip("-")
    return s or "unknown"


def get_filename(entry):
    """Derive filename: {category}-{number}-{name}.json (all lowercase, no spaces)."""
    category = sanitize(entry.get("category"))
    number = entry.get("number")
    num = (number.lower().replace("/", "-").replace("\\", "-") if number else "unknown")
    name = sanitize(entry.get("name"))
    return f"{category}-{num}-{name}.json"


def main():
    if not SOURCE.exists():
        print("Source file not found:", SOURCE)
        raise SystemExit(1)

    data = json.loads(SOURCE.read_text(encoding="utf-8"))
    if not isinstance(data, list):
        print("Expected JSON array")
        raise SystemExit(1)

    seen = {}
    skipped = 0

    for i, entry in enumerate(data):
        type_val = entry.get("type")
        category = entry.get("category")
        number = entry.get("number")
        name = entry.get("name")

        if not type_val or not category or not number or not name:
            print(f"[{i}] Skipping entry missing type/category/number/name: {number or name or '(no id)'}")
            skipped += 1
            continue

        type_folder = type_to_folder(type_val)
        dir_path = SRC_DIR / type_folder
        dir_path.mkdir(parents=True, exist_ok=True)

        filename = get_filename(entry)
        key = str(dir_path / filename)
        count = seen.get(key, 0)
        seen[key] = count + 1
        if count > 0:
            base, ext = filename.rsplit(".", 1)
            filename = f"{base}-{count + 1}.{ext}"

        out_path = dir_path / filename
        out_path.write_text(json.dumps(entry, indent=4, ensure_ascii=False), encoding="utf-8")

    print(f"Wrote {len(data) - skipped} entries to {SRC_DIR}")
    if skipped:
        print(f"Skipped {skipped} entries.")


if __name__ == "__main__":
    main()
